libff/algebra/scalar_multiplication/multiexp.tcc

```diff
@@ -27,6 +27,246 @@
+    {
+        const ssize_t digit = fixed_wnaf_digit(exponents[i], c, digit_idx);
+        if (digit == 0)
+        {
+            continue;
+        }
+
+        // Unroll each branch, to avoid copying the group elements.
+        if (digit < 0)
+        {
+            const size_t bucket_idx = (-digit) - 1;
+            assert(bucket_idx < num_buckets);
+            if (bucket_hit[bucket_idx])
+            {
+                if (MixedAddition)
+                {
+                    buckets[bucket_idx] = buckets[bucket_idx].mixed_add(- bases[i]);
+                }
+                else
+                {
+                    buckets[bucket_idx] = buckets[bucket_idx].add(- bases[i]);
+                }
+            }
+            else
+            {
+                buckets[bucket_idx] = - bases[i];
+                bucket_hit[bucket_idx] = true;
+                ++num_buckets_initialized;
+            }
+        }
+        else
+        {
+            const size_t bucket_idx = digit - 1;
+            assert(bucket_idx < num_buckets);
+            if (bucket_hit[bucket_idx])
+            {
+                if (MixedAddition)
+                {
+                    buckets[bucket_idx] = buckets[bucket_idx].mixed_add(bases[i]);
+                }
+                else
+                {
+                    buckets[bucket_idx] = buckets[bucket_idx].add(bases[i]);
+                }
+            }
+            else
+            {
+                buckets[bucket_idx] = bases[i];
+                bucket_hit[bucket_idx] = true;
+                ++num_buckets_initialized;
+            }
+
+        }
+    }
+
+    // Check up-front for the edge-case where no buckets have been touched.
+    if (num_buckets_initialized == 0)
+    {
+        return GroupT::zero();
+    }
+
+    return multiexp_accumulate_buckets(buckets, bucket_hit, num_buckets);
+}
+
+template<typename GroupT, typename FieldT, bool MixedAddition>
+GroupT multiexp_signed_digits(
+    typename std::vector<GroupT>::const_iterator bases,
+    typename std::vector<GroupT>::const_iterator bases_end,
+    typename std::vector<FieldT>::const_iterator exponents,
+    typename std::vector<FieldT>::const_iterator exponents_end)
+{
+    UNUSED(exponents_end);
+
+    const size_t num_entries = bases_end - bases;
+    assert(exponents_end - exponents == (ssize_t)num_entries);
+    const size_t c = pippenger_optimal_c(num_entries) + 1;
+    assert(c > 0);
+    using bigint_t =
+        typename std::decay<decltype(((FieldT*)nullptr)->mont_repr)>::type;
+
+    // Pre-compute the bigint values
+    size_t num_bits = 0;
+    std::vector<bigint_t> bi_exponents(num_entries);
+    for (size_t i = 0 ; i < num_entries ; ++i)
+    {
+        bi_exponents[i] = exponents[i].as_bigint();
+        num_bits = std::max(num_bits, bi_exponents[i].num_bits());
+    }
+
+    const size_t num_rounds = (num_bits + c - 1) / c;
+
+    // Digits have values v -2^{c-1} \leq v \lt 2^{c-1}, i.e. 0 \leq |v| \leq
```

Comment:
`Digits have values v -2^{c-1} \leq v \lt 2^{c-1}` -> `Digits have values v, s.t. -2^{c-1} \leq v \lt 2^{c-1}`

Reply:
`0 \leq |v|` -> this leg of the inequality is not really needed (we know that the absolute value maps to \RR^+)

Reply:
I've reworded this comment to clarify.